File: src/lib/paf/scoring.ts
// Stage classification, conversion ratios, vs-Ethereum scoring, trajectory.
// Port of paf-toolkit/paf/scoring.py.

import type {
  ConversionRatios,
  HistoricalSnapshot,
  RawPafSnapshot,
  Stage,
} from "./types"
import { TRANSITIONS } from "./thresholds"

// ─── Stage classification ───────────────────────────────────────────────────

export function classifyStage(
  s: RawPafSnapshot,
  nativeLiquidStakingFlag = false,
): Stage {
  // Native LS networks (Solana, Cardano) classify as S* once they cross
  // the 2% staking-ratio floor — the funnel L3 evaluation does not apply.
  if (nativeLiquidStakingFlag) {
    return s.stakingRatioPct <= 2.0 ? "S0" : "S*"
  }

  let current: Stage = "S0"
  for (const t of TRANSITIONS) {
    if (t.fromStage !== current) continue
    if (t.check(s)) {
      current = t.toStage
    } else {
      break
    }
  }
  return current
}

// ─── Conversion ratios ──────────────────────────────────────────────────────

export function computeConversionRatios(s: RawPafSnapshot): ConversionRatios {
  const ratios: ConversionRatios = {
    stakingParticipationPct: s.stakingRatioPct,
    liquidizationPct: s.liquidizationRatePct,
    defiProductivityPct: s.defiProductivityRatePct,
    restakingPct: null,
    loopingDepthPct: null,
  }

  if (s.restakingTvlUsd != null && s.lstTvlUsd != null && s.lstTvlUsd > 0) {
    ratios.restakingPct = (s.restakingTvlUsd / s.lstTvlUsd) * 100
  }

  if (
    s.pendleEquivalentTvlUsd != null &&
    s.lstTvlUsd != null &&
    s.lstTvlUsd > 0
  ) {
    ratios.loopingDepthPct = (s.pendleEquivalentTvlUsd / s.lstTvlUsd) * 100
  }

  return ratios
}

// ─── vs-Ethereum scoring (log-scaled, 100 = parity, cap 150) ────────────────

function logScaleScore(networkValue: number, ethValue: number, cap = 150): number {
  if (ethValue <= 0 || networkValue <= 0) return 0
  const raw = 100 * (1 + Math.log10(networkValue / ethValue))
  return Math.max(0, Math.min(cap, raw))
}

export type VsEthScores = {
  L1: number | null
  L2: number | null
  L3: number | null
  "L4.1": number | null
  "L4.2": number | null
  overall: number
}

export function scoreVsEthereum(network: RawPafSnapshot, eth: RawPafSnapshot): VsEthScores {
  const scores = {
    L1: logScaleScore(network.marketCapUsd, eth.marketCapUsd),
    L2: logScaleScore(network.stakingRatioPct, eth.stakingRatioPct),
    L3:
      network.liquidizationRatePct != null && eth.liquidizationRatePct != null
        ? logScaleScore(network.liquidizationRatePct, eth.liquidizationRatePct)
        : null,
    "L4.1":
      network.restakingRatePct != null && eth.restakingRatePct != null
        ? logScaleScore(network.restakingRatePct, eth.restakingRatePct)
        : null,
    "L4.2":
      network.defiProductivityRatePct != null &&
      eth.defiProductivityRatePct != null
        ? logScaleScore(
            network.defiProductivityRatePct,
            eth.defiProductivityRatePct,
          )
        : null,
  } as const

  const values = Object.values(scores).filter((v): v is number => v != null)
  const overall = values.length > 0 ? values.reduce((a, b) => a + b, 0) / values.length : 0

  return { ...scores, overall }
}

// ─── Trajectory positioning ─────────────────────────────────────────────────

function snapshotDistance(s: RawPafSnapshot, h: HistoricalSnapshot): number {
  const diffs: number[] = []

  diffs.push(Math.abs(s.stakingRatioPct - h.stakingRatioPct))

  if (s.liquidizationRatePct != null && h.liquidizationRatePct != null) {
    diffs.push(Math.abs(s.liquidizationRatePct - h.liquidizationRatePct))
  }

  if (s.defiProductivityRatePct != null && h.defiProductivityRatePct != null) {
    diffs.push(
      Math.abs(s.defiProductivityRatePct - h.defiProductivityRatePct),
    )
  }

  if (diffs.length === 0) return Infinity
  return diffs.reduce((a, b) => a + b, 0) / diffs.length
}

export function trajectoryPosition(
  s: RawPafSnapshot,
  history: HistoricalSnapshot[],
): { description: string; nearestQuarter: string | null } {
  if (history.length === 0) {
    return { description: "no benchmark history available", nearestQuarter: null }
  }

  let nearest = history[0]
  let bestDist = snapshotDistance(s, nearest)
  for (const h of history.slice(1)) {
    const d = snapshotDistance(s, h)
    if (d < bestDist) {
      bestDist = d
      nearest = h
    }
  }

  const suffix = nearest.note ? ` — ${nearest.note}` : ""
  return {
    description: `comparable to ETH ${nearest.quarter}${suffix}`,
    nearestQuarter: nearest.quarter,
  }
}

// ─── BD recommendation ──────────────────────────────────────────────────────

export function bdRecommendation(stage: Stage): string {
  switch (stage) {
    case "S0":
      return "Watch list, not active target. Network has not crossed the 2% staking-ratio floor. Revisit if a staking-incentive program or consensus-upgrade catalyst is announced."
    case "S1":
      return "Highest-leverage BD opportunity. Network is securing but pre-liquidization. Pitch a liquid-staking product before a dominant issuer emerges."
    case "S2":
      return "Active BD target. LST market is forming. Push for DeFi integrations (lending + AMM) to break the 30% productivity barrier."
    case "S3":
      return "Productive integrations stage. Restaking primitives are the next gate but watch for restaking market stability before pitching restaking-only narratives."
    case "S4.1":
      return "Restaking-era opportunity with caution. Pitch composability primitives — yield tokenization, looping markets, LRT-collateralized lending."
    case "S4.2":
      return "Mature funnel. All conversion ratios qualify for category-leadership pitches. BD focus on differentiated yield products and institutional structured products."
    case "S*":
      return "Native liquid staking — non-standard funnel. Focus on L4.2 productivity and L4.1 restaking analogs; L3 evaluation does not apply."
    default:
      return "No specific recommendation."
  }
}

// ─── Strengths & gaps (auto-derived from threshold proximity) ───────────────

export function strengthsAndGaps(
  s: RawPafSnapshot,
  eth: RawPafSnapshot,
  stage: Stage,
): { strengths: string[]; gaps: string[] } {
  const strengths: string[] = []
  const gaps: string[] = []

  if (s.stakingRatioPct >= eth.stakingRatioPct) {
    strengths.push(
      `Staking ratio ${s.stakingRatioPct.toFixed(1)}% meets/exceeds Ethereum's ${eth.stakingRatioPct.toFixed(1)}%.`,
    )
  } else if (s.stakingRatioPct < 0.5 * eth.stakingRatioPct) {
    gaps.push(
      `Staking ratio ${s.stakingRatioPct.toFixed(1)}% is less than half Ethereum's — security thin.`,
    )
  }

  if (s.largestLstIssuerSharePct != null && s.largestLstIssuerSharePct >= 70) {
    gaps.push(
      `LST market dominated by single issuer (${s.largestLstIssuerSharePct.toFixed(0)}% share) — anti-monopoly red flag.`,
    )
  }

  if (s.liquidizationRatePct != null && s.liquidizationRatePct >= 30) {
    strengths.push(
      `Liquidization rate ${s.liquidizationRatePct.toFixed(0)}% is healthy (S2-mature).`,
    )
  }

  if (s.nativeRedemptionAvailable === false) {
    gaps.push("Native LST redemption not yet enabled — S2→S3 blocker.")
  }

  if (s.defiProductivityRatePct != null) {
    if (s.defiProductivityRatePct >= 30) {
      strengths.push(
        `DeFi productivity ${s.defiProductivityRatePct.toFixed(0)}% clears the S3 threshold.`,
      )
    } else if (s.defiProductivityRatePct < 10) {
      gaps.push(
        `DeFi productivity ${s.defiProductivityRatePct.toFixed(0)}% is low — LST is not yet productive collateral.`,
      )
    }
  }

  const nextT = TRANSITIONS.find((t) => t.fromStage === stage)
  if (nextT) {
    gaps.push(`Next gate: ${nextT.fromStage} → ${nextT.toStage} requires ${nextT.rule}.`)
  }

  return { strengths, gaps }
}

File: paf-toolkit/paf/scoring.py
"""Stage classification, conversion ratios, vs-Ethereum scoring, trajectory."""

from __future__ import annotations

import math
from typing import Optional

from .schema import (
    EthereumBenchmark,
    HistoricalSnapshot,
    NetworkProfile,
    Stage,
)
from .thresholds import TRANSITIONS


# ---------------------------------------------------------------------------
# Stage classification
# ---------------------------------------------------------------------------

def classify_stage(profile: NetworkProfile) -> Stage:
    """Walk the threshold ladder from S0 upward, returning the highest stage met.

    Native-liquid-staking networks return S_STAR if they pass L1+L2 thresholds
    but the L3 evaluation is skipped by design (their staked positions are
    already liquid by protocol construction).
    """
    if profile.native_liquid_staking_flag:
        # Native LS networks still need a minimum stake ratio to count as "secured".
        if profile.l2.staking_ratio_pct.value <= 2.0:
            return Stage.S0
        return Stage.S_STAR

    current = Stage.S0
    for t in TRANSITIONS:
        if t.from_stage != current:
            continue
        if t.check(profile):
            current = t.to_stage
        else:
            break
    return current


# ---------------------------------------------------------------------------
# Conversion ratios
# ---------------------------------------------------------------------------

def compute_conversion_ratios(profile: NetworkProfile) -> dict[str, Optional[float]]:
    """Return the five funnel conversion ratios (all as percentages or None).

    Keys:
      - staking_participation_pct    (L1 → L2)
      - liquidization_pct            (L2 → L3)
      - defi_productivity_pct        (L3 → L4.2)
      - restaking_pct                (L3 → L4.1)
      - looping_depth_pct            (heuristic: pendle_tvl / lst_tvl)
    """
    ratios: dict[str, Optional[float]] = {
        "staking_participation_pct": profile.l2.staking_ratio_pct.value,
        "liquidization_pct": None,
        "defi_productivity_pct": None,
        "restaking_pct": None,
        "looping_depth_pct": None,
    }

    if profile.l3 is not None:
        ratios["liquidization_pct"] = profile.l3.liquidization_rate_pct.value

    if profile.l42 is not None:
        ratios["defi_productivity_pct"] = profile.l42.defi_productivity_rate_pct.value

    if profile.l41 is not None and profile.l3 is not None:
        lst_tvl = profile.l3.lst_tvl_usd.value
        if lst_tvl > 0:
            ratios["restaking_pct"] = (
                profile.l41.restaking_tvl_usd.value / lst_tvl * 100.0
            )

    if profile.l42 is not None and profile.l3 is not None:
        lst_tvl = profile.l3.lst_tvl_usd.value
        if lst_tvl > 0:
            ratios["looping_depth_pct"] = (
                profile.l42.pendle_equivalent_tvl_usd.value / lst_tvl * 100.0
            )

    return ratios


# ---------------------------------------------------------------------------
# vs-Ethereum scoring
# ---------------------------------------------------------------------------

def _log_scale_score(network_value: float, eth_value: float, cap: float = 150.0) -> float:
    """Score on a log scale: 100 = parity with Ethereum, capped at `cap`.

    score = 100 * (1 + log10(network / eth))  clamped to [0, cap]
    so:
      - network = eth        → 100
      - network = 10x eth    → 200 → capped to 150
      - network = 0.1x eth   → 0
      - network = 0.5x eth   → ~70
    """
    if eth_value <= 0 or network_value <= 0:
        return 0.0
    raw = 100.0 * (1.0 + math.log10(network_value / eth_value))
    return max(0.0, min(cap, raw))


def score_vs_ethereum(
    profile: NetworkProfile,
    eth_benchmark: EthereumBenchmark,
) -> dict[str, float]:
    """Per-layer 0–150 scores comparing the network to Ethereum's current snapshot.

    Each layer score is the score of its single most-load-bearing KPI
    (chosen because it's the one that drives funnel-stage progression):

      L1   → market_cap_usd                       (token adoption depth)
      L2   → staking_ratio_pct                    (security participation)
      L3   → liquidization_rate_pct               (LST adoption depth)
      L4.1 → restaking_rate_pct                   (restaking depth)
      L4.2 → defi_productivity_rate_pct           (productive deployment)
    """
    eth = eth_benchmark.snapshot
    scores: dict[str, float] = {}

    scores["L1"] = _log_scale_score(
        profile.l1.market_cap_usd.value,
        eth.l1.market_cap_usd.value,
    )
    scores["L2"] = _log_scale_score(
        profile.l2.staking_ratio_pct.value,
        eth.l2.staking_ratio_pct.value,
    )
    if profile.l3 is not None and eth.l3 is not None:
        scores["L3"] = _log_scale_score(
            profile.l3.liquidization_rate_pct.value,
            eth.l3.liquidization_rate_pct.value,
        )
    else:
        scores["L3"] = 0.0

    if profile.l41 is not None and eth.l41 is not None:
        scores["L4.1"] = _log_scale_score(
            profile.l41.restaking_rate_pct.value,
            eth.l41.restaking_rate_pct.value,
        )
    else:
        scores["L4.1"] = 0.0

    if profile.l42 is not None and eth.l42 is not None:
        scores["L4.2"] = _log_scale_score(
            profile.l42.defi_productivity_rate_pct.value,
            eth.l42.defi_productivity_rate_pct.value,
        )
    else:
        scores["L4.2"] = 0.0

    scores["overall"] = sum(scores.values()) / len(scores)
    return scores


# ---------------------------------------------------------------------------
# Trajectory positioning
# ---------------------------------------------------------------------------

def _snapshot_distance(profile: NetworkProfile, snap: HistoricalSnapshot) -> float:
    """Distance between a profile and a historical ETH snapshot.

    We compare on three normalized dimensions (staking, liquidization, defi
    productivity), each clipped to [0, 100]. Missing dimensions are skipped;
    distance is mean absolute difference of the available pairs.
    """
    diffs: list[float] = []

    diffs.append(abs(profile.l2.staking_ratio_pct.value - snap.staking_ratio_pct))

    if profile.l3 is not None and snap.liquidization_rate_pct is not None:
        diffs.append(
            abs(profile.l3.liquidization_rate_pct.value - snap.liquidization_rate_pct)
        )

    if profile.l42 is not None and snap.defi_productivity_rate_pct is not None:
        diffs.append(
            abs(
                profile.l42.defi_productivity_rate_pct.value
                - snap.defi_productivity_rate_pct
            )
        )

    return sum(diffs) / len(diffs) if diffs else float("inf")


def trajectory_position(
    profile: NetworkProfile,
    eth_history: list[HistoricalSnapshot],
) -> str:
    """Find the ETH quarter whose funnel state most resembles this profile.

    Returns a short human-readable string like:
        "comparable to ETH 2022-Q3 (Merge era)"
    """
    if not eth_history:
        return "no benchmark history available"

    nearest = min(eth_history, key=lambda s: _snapshot_distance(profile, s))
    suffix = f" — {nearest.note}" if nearest.note else ""
    return f"comparable to ETH {nearest.quarter}{suffix}"
